### PR TITLE
chore(`vm`): rewrite `VmExecutionResult` as `Result`

### PR DESCRIPTION
**Pull request by EdwardJES**
This PR addresses https://github.com/risechain/pevm/issues/318

### CODE REVIEW

File: src/vm.rs [hunk 1]
```diff
@@ -518,8 +519,11 @@ impl<'a, S: Storage, C: PevmChain> Vm<'a, S, C> {
         // Execute
         let mut db = match VmDb::new(self, tx_version.tx_idx, tx, from_hash, to_hash) {
             Ok(db) => db,
-            // TODO: Handle different errors differently
-            Err(_) => return VmExecutionResult::FallbackToSequential,
+            Err(err) => {
+                return VmExecutionResult::Err(VmExecutionError::ExecutionError(
```
Comment (EdwardJES) [inline src/vm.rs:523]: Hey @hai-rise, I'm not the biggest fan of the stuttering and verbosity here. 

An alternative could impl method e.g.

```
impl VmExecutionError {
    pub fn execution_error(error: ExecutionError) -> Self {
        VmExecutionError::ExecutionError(error)
    }
}
```
Please lmk what you think 🙏

Comment (hai-rise) [reply src/vm.rs:523]: I agree! I think a clean solution is to:
- Bring `ReadError` from `lib.rs` here as a `pub(crate)` (as a related chore).
- Implement a `From<ReadError>`  for `VmExecutionError` as the source-of-truth for read error handling behaviour.
```rust
impl From<ReadError> for VmExecutionError {
    fn from(err: ReadError) -> Self {
        match err {
            ReadError::InconsistentRead => VmExecutionError::Retry,
            ReadError::SelfDestructedAccount => VmExecutionError::FallbackToSequential,
            ReadError::Blocking(tx_idx) => VmExecutionError::Blocking(tx_idx),
            _ => VmExecutionError::ExecutionError(EVMError::Database(err)),
        }
    }
}
```
- Then we can simplify this instance to:
```rust
let mut db = VmDb::new(self, tx_version.tx_idx, tx, from_hash, to_hash)
            .map_err(VmExecutionError::from)?;
```
- And the several match arms below to:
```rust
Err(EVMError::Database(read_error)) => Err(read_error.into()),
```

Also, note that the current change is a semantic change (from returning `FallbackToSequential` to returning `ExecutionError`) that fails test.

File: src/vm.rs [hunk 2]
```diff
@@ -70,20 +71,20 @@ impl PevmTxExecutionResult {
...
-    },
+    ReadError { blocking_tx_idx: TxIdx },
     ExecutionError(ExecutionError),
-    Ok {
-        execution_result: PevmTxExecutionResult,
-        flags: FinishExecFlags,
-    },
 }
 
+pub(crate) struct VmExecutionOutcome {
```
Comment (hai-rise) [inline src/vm.rs:81]: Naming here is hard! We can keep `VmExecutionResult` for now.

File: src/vm.rs [hunk 3]
```diff
@@ -70,20 +71,20 @@ impl PevmTxExecutionResult {
...
+    pub execution_result: PevmTxExecutionResult,
```
Comment (hai-rise) [inline src/vm.rs:82]: Let's do `pub(crate)` here.

File: src/vm.rs [hunk 4]
```diff
@@ -70,20 +71,20 @@ impl PevmTxExecutionResult {
...
+    pub flags: FinishExecFlags,
+}
+
+pub(crate) type VmExecutionResult = Result<VmExecutionOutcome, VmExecutionError>;
```
Comment (hai-rise) [inline src/vm.rs:86]: Inline `Result<VmExecutionResult, VmExecutionError>` to the signature of `Vm::execute` is cleaner.

File: src/vm.rs [hunk 5]
```diff
@@ -646,21 +650,25 @@ impl<'a, S: Storage, C: PevmChain> Vm<'a, S, C> {
                     flags |= FinishExecFlags::WroteNewLocation;
                 }
 
-                VmExecutionResult::Ok {
+                let outcome = VmExecutionOutcome {
```
Comment (hai-rise) [inline src/vm.rs:653]: Just wrap this in an `Ok`.

**Review by hai-rise — CHANGES_REQUESTED** [review]
I've left a few suggestions. Thank you again for the work!

Also, it'd be nice if this PR also adds `Result<(), VmExecutionError>` to `apply_rewards`, to remove the `unreachable!` panic to: 
```rust
_ => return Err(ReadError::InvalidMemoryValueType.into()),
```

🙏

**Comment by EdwardJES** [comment]
Thank you for the quick feedback @hai-rise, this is ready for another review 🙏

File: src/vm.rs [hunk 6]
```diff
@@ -70,18 +71,51 @@ impl PevmTxExecutionResult {
...
+    /// TODO: Handle this at the type level?
+    InvalidMemoryValueType,
+}
+
+impl From<ReadError> for VmExecutionError {
+    fn from(err: ReadError) -> Self {
+        match err {
+            ReadError::InconsistentRead => VmExecutionError::Retry,
+            ReadError::SelfDestructedAccount => VmExecutionError::FallbackToSequential,
+            ReadError::BlockingIndex(tx_idx) => VmExecutionError::Blocking {
```
Comment (hai-rise) [inline src/vm.rs:108]: Very minor but we can unify both as `Blocking(TxIdx)` for consistency now 🙏.

**Review by hai-rise — CHANGES_REQUESTED** [review]
Just one minor change left, otherwise it's perfect. Thank you!

**Comment by EdwardJES** [comment]
Thank you @hai-rise, unified the change. I'm looking forward to making more contributions.

**Comment by EdwardJES** [comment]
Thank you @hai-rise, unified the change. I'm looking forward to making more contributions.

File: src/vm.rs [hunk 7]
```diff
@@ -70,18 +71,51 @@ impl PevmTxExecutionResult {
...
 
-// TODO: Rewrite as [Result]
-pub(crate) enum VmExecutionResult {
+pub(crate) enum VmExecutionError {
     Retry,
     FallbackToSequential,
-    ReadError {
-        blocking_tx_idx: TxIdx,
-    },
+    Blocking { blocking_tx_idx: TxIdx },
```
Comment (hai-rise) [inline src/vm.rs:77]: Sorry, I should have been clearer. But we should also change this to `Blocking(TxIdx)`, so no more `blocking_tx_idx` for both error types.

**Review by hai-rise — CHANGES_REQUESTED** [review]

**Comment by EdwardJES** [comment]
No worries, @hai-rise. I probably should've caught that list nit.

**Review by hai-rise — APPROVED** [review]
Better structure and error handling, thank you!

**Pull request merged**